rust/kernel/types.rs

```diff
@@ -49,46 +61,121 @@ pub trait PointerWrapper {
 }
 
 impl<T> PointerWrapper for Box<T> {
+    type Borrowed = UnsafeReference<T>;
+
     fn into_pointer(self) -> *const c_types::c_void {
         Box::into_raw(self) as _
     }
 
+    unsafe fn borrow(ptr: *const c_types::c_void) -> Self::Borrowed {
+        // SAFETY: The safety requirements for this function ensure that the object is still alive,
+        // so it is safe to dereference the raw pointer.
+        // The safety requirements also ensure that the object remains alive for the lifetime of
+        // the returned value.
+        unsafe { UnsafeReference::new(&*ptr.cast()) }
+    }
+
     unsafe fn from_pointer(ptr: *const c_types::c_void) -> Self {
         // SAFETY: The passed pointer comes from a previous call to [`Self::into_pointer()`].
         unsafe { Box::from_raw(ptr as _) }
     }
 }
 
 impl<T: RefCounted> PointerWrapper for Ref<T> {
+    type Borrowed = UnsafeReference<T>;
+
     fn into_pointer(self) -> *const c_types::c_void {
         Ref::into_raw(self) as _
     }
 
+    unsafe fn borrow(ptr: *const c_types::c_void) -> Self::Borrowed {
+        // SAFETY: The safety requirements for this function ensure that the object is still alive,
+        // so it is safe to dereference the raw pointer.
+        // The safety requirements also ensure that the object remains alive for the lifetime of
+        // the returned value.
+        unsafe { UnsafeReference::new(&*(ptr as *const T)) }
```

Comment:
Nit: left out a `cast()` here?

Reply:
Oops. I had missed this one because it goes away when I update `Ref`'s implementation.

Anyway, fixed it now. PTAL.